verilog/formatting/align.cc

```diff
@@ -489,6 +481,139 @@ class DataDeclarationColumnSchemaScanner : public ColumnSchemaScanner {
 };
 
+class ParameterDeclarationColumnSchemaScanner : public ColumnSchemaScanner {
+ public:
+  ParameterDeclarationColumnSchemaScanner() = default;
+
+  void Visit(const SyntaxTreeNode& node) override {
+    auto tag = NodeEnum(node.Tag().tag);
+    VLOG(2) << __FUNCTION__ << ", node: " << tag << " at "
+            << TreePathFormatter(Path());
+    if (new_column_after_open_bracket_) {
+      ReserveNewColumn(node, FlushRight);
+      new_column_after_open_bracket_ = false;
+      TreeContextPathVisitor::Visit(node);
+      return;
+    }
+    
+    switch (tag) {
+      case NodeEnum::kTypeInfo: {
+        SyntaxTreePath new_path{1};
+        const ValueSaver<SyntaxTreePath> path_saver(&current_path_, new_path);
+        ReserveNewColumn(node, FlushLeft);
+        break;
+      }
+      
+      case NodeEnum::kTrailingAssign: {
+        ReserveNewColumn(node, FlushLeft);
+        break;
+      }
+
+      case NodeEnum::kUnqualifiedId: {
+        if(Context().DirectParentIs(NodeEnum::kParamType)) {
+          ReserveNewColumn(node, FlushLeft);
+        }
+        break;
+      }
+      default:
+        break;
+    }
+
+    // recursive visitation
+    TreeContextPathVisitor::Visit(node);
+    VLOG(2) << __FUNCTION__ << ", leaving node: " << tag;
+  }
+
+  void Visit(const SyntaxTreeLeaf& leaf) override {
+    VLOG(2) << __FUNCTION__ << ", leaf: " << leaf.get() << " at "
+            << TreePathFormatter(Path());
+
+    if (new_column_after_open_bracket_) {
+      ReserveNewColumn(leaf, FlushRight);
+      new_column_after_open_bracket_ = false;
+      return;
+    }
+
+    const int tag = leaf.get().token_enum();
+    switch (tag) {
+      // Align keywords 'parameter', 'localparam' and 'type' under the same column.
+      case verilog_tokentype::TK_parameter:
+      case verilog_tokentype::TK_localparam: {
+        ReserveNewColumn(leaf, FlushLeft);
+        break;
+      }
+
+      case verilog_tokentype::TK_type: {
+        if(Context().DirectParentIs(NodeEnum::kParamDeclaration)) {
+          ReserveNewColumn(leaf, FlushLeft);
+        }
+        break;
+      }
+
+      // Sometimes the parameter indentifier which is of token SymbolIdentifier can
+      // appear at different paths depending on the parameter type. Make them aligned 
+      // so they fall under the same column.
+      case verilog_tokentype::SymbolIdentifier: {
+         if(current_path_ == SyntaxTreePath{2, 0}) {
+          SyntaxTreePath new_path{1, 2};
+          const ValueSaver<SyntaxTreePath> path_saver(&current_path_, new_path);
+          ReserveNewColumn(leaf, FlushLeft);
+          return;
+         }
+         
+         if(Context().DirectParentIs(NodeEnum::kParamType)) 
+           ReserveNewColumn(leaf, FlushLeft);
+        break;
+      }
+
+      // '=' is another column where things should be aligned. But type declarations 
+      // and localparam cause '=' to appear under two different paths in CST. Align
+      // them.
```

Comment:
Yeah, I wonder if these can be restructured more consistently...
No changed needed here.